### ISSUE TITLE
Cannot launch using latest client files

### ISSUE THREAD

**Issue by ptgeorge** [VoxelBoy/MobileUO#24]
Trying to launch using the latest client files but I receive an error NullReferenceException related to ReplaceUopGroup.  Please see the screenshot for the full error.
![IMG_0146](https://github.com/user-attachments/assets/b06ad114-41e4-4bf3-8690-bfd13c1a8415)

**Comment by Leigheas** [comment]
Yep the changes to classicuo to allow for the new files to be read are not present in mobileuo. So until someone updates mobileuo to basically the newest classicuo files the. 103 files are the last ones it can read

**Comment by ptgeorge** [comment]
Thanks for the info.  Any idea what the level of effort is to update from the latest ClassicUO?  It looks like this repo wasn't forked and subtree wasn't used, so its fairly difficult to see what changes to the CUO code were made, if any.

**Comment by ptgeorge** [comment]
Thanks.  This is really helpful.  I’ll take a look at some point and see what kind of drift there is.  I might be able to take it on.

**Comment by ptgeorge** [comment]
It looks like this is the closest CUO commit: https://github.com/ClassicUO/ClassicUO/tree/303c2ae699ca02caa6b09fbe5fbcb0c92f8bbd68

Doing a diff and reviewing looks to be changes made for MobileUO that aren't in CUO.  A lot can happen in 4 years 😁 There's definitely been a lot of restructuring in CUO so as you said, @VoxelBoy , its a lot to bring MobileUO up to the latest.

**Comment by ptgeorge** [comment]
@VoxelBoy I’ve opened #25 to make it easier for me to identify the MobileUO modifications to CUO code.  Feel free to review, but it’s definitely not urgent.  I just didn’t want to include that sort of content in a PR while actually making updates of CUO.

**Comment by VoxelBoy** [comment]
Very nice @ptgeorge. I've merged your PR. A quick look at the diff seems to show that a lot of the changes can be minimized between the two versions of the ClassicUO src. Might be a good first step before attempting to update ClassicUO to latest.

**Comment by ptgeorge** [comment]
Thanks!  I’ll see if I can reduce the changes made to CUO within MobileUO.  I’ve already been going down a few different routes to see how to integrate the latest changes.  It’s some work but eventually I’ll probably work it out.

To fully modernize, the Assistant work within MobileUO will likely be the most difficult, as “Packet” no longer exists within CUO.

**Comment by ptgeorge** [comment]
@VoxelBoy Do you know what the original source was for XNAEmulator?  Was it https://github.com/mvi/UnityXNA/ or some other repo?

**Comment by VoxelBoy** [comment]
@ptgeorge I didn't pull it in from a public repo, a kind person (whose handle I don't remember atm) on the ClassicUO Discord gave me his implementation at the time, which could have been based on a public repo like the one you linked at but I don't know for sure..

**Comment by mandlar** [comment]
> Thanks! I’ll see if I can reduce the changes made to CUO within MobileUO. I’ve already been going down a few different routes to see how to integrate the latest changes. It’s some work but eventually I’ll probably work it out.
> 
> To fully modernize, the Assistant work within MobileUO will likely be the most difficult, as “Packet” no longer exists within CUO.

What are you thinking for the best way to try and bring MobileUO up to the latest version of CUO? A big bang of 4 years of updates looks so unwieldy to me.

**Comment by VoxelBoy** [comment]
Probably by updating the classicuo copy inside mobileuo from one large refactor to another while making sure it's still working.

**Comment by mandlar** [comment]
To help familiarize myself with the project and see the scope of changes between MobileUO and ClassicUO, I made a list of all the differences between the two projects here: https://github.com/mandlar/MobileUO/wiki/MobileUO-changes-from-ClassicUO

As next steps, I'll try and reduce some of these changes before trying to roll forward with ClassicUO updates in chunks.

**Comment by ptgeorge** [comment]
Nice.  FYI, I did start some various experiments in my fork: https://github.com/ptgeorge/MobileUO

Not sure if my latest is up there or not.  It’s been a couple of weeks since I’ve looked at it.  The assistant stuff probably needs to be temporarily stripped away just to at least get a minimal modern build up.

“diff -r” with various filtering is what I was using to compare.

**Comment by mandlar** [comment]
I opened a PR to further reduce changes from ClassicUO and symlink some more files.

https://github.com/VoxelBoy/MobileUO/pull/26

**Comment by ptgeorge** [comment]
I did a thing.  I rebuilt mobile support based upon ClassicUO from a couple weeks ago.  There’s plenty of bugs and lots of features that need to be added.  This was more of an experiment than something I can continually dedicate time to.

https://gitlab.com/ptgeorge-open/MobUO

READMEs:
- [Android](https://gitlab.com/ptgeorge-open/MobUO/-/blob/main/src/ClassicUO.iOS/README.md)
- [iOS](https://gitlab.com/ptgeorge-open/MobUO/-/blob/main/src/ClassicUO.iOS/README.md)

Mirrored to GitHub: https://github.com/ptg-open/MobUO

**Comment by mandlar** [comment]
@ptgeorge in case you weren't aware, MobileUO has moved to here: https://github.com/MobileUO/MobileUO

In my dev branch, it's the latest CUO (as of version 1.1.0.178, so it's missing maybe the last year or so). I also just pushed a PR this week that fixed a major file IO performance issue from CUO that was killing our performance on Mobile devices.

If you want to PR anything upstream, I more than welcome them!

**Comment by ptgeorge** [comment]
Nice @mandlar — had no idea about the move or that you were successful in getting it updated to the latest.  Unity support for newer C# versions seemed to be lacking when I tried to update this repo with the latest.

Thanks for working on this!

**Comment by mandlar** [comment]
@ptgeorge yes, CUO uses a lot of C# 9 language features that aren't available in Unity. If you pull my dev branch and search for "MobileUO:" comments, you'll see my workarounds for all of those.

**Comment by ptgeorge** [comment]
Interesting.  Take a look at https://gitlab.com/ptgeorge-open/MobUO/-/compare/0a1c53dd1e928a276e1926a052f8ba4f4c3ddaaf...main?from_project_id=80509413 (or https://github.com/ptg-open/MobUO/compare/0a1c53dd1e928a276e1926a052f8ba4f4c3ddaaf..main ) to see the changes.

.NET 10 is used in my version.  

Most of the changes to the upstream CUO code is to allow cleaning up, so games can be relaunched on iOS / Android.

In theory, it should be easy to pull in changes from upstream CUO with my version.
